manim/utils/hashing.py

```diff
@@ -0,0 +1,150 @@
+import json
+import zlib
+import inspect
+import copy
+import dis
+import numpy as np
+from types import ModuleType
+
+from ..logger import logger
+
+
+class CustomEncoder(json.JSONEncoder):
+    def default(self, obj):
+        """
+        This method is used to serialize objects to JSON format.
+        If obj is a function, then it will return a dict with two keys : 'code', for the code source, and 'nonlocals' for all nonlocalsvalues. (including nonlocals functions, that will be serialized as this is recursive.)
+        if obj is a np.darray, it converts it into a list.
+        if obj is an object with __dict__ attribute, it returns its __dict__.
+        Else, will let the JSONEncoder do the stuff, and throw an error if the type is not suitable for JSONEncoder. 
+
+        Parameters
+        ----------
+        obj : `any`
+            Arbitrary object to convert
+
+        Returns
+        -------
+        `any`
+            Python object that JSON encoder will recognize
+
+        """
+        if inspect.isfunction(obj) and not isinstance(obj, ModuleType):
+            r = inspect.getclosurevars(obj)
+            x = {**copy.copy(r.globals), **copy.copy(r.nonlocals)}
+            for i in list(x):
+                # NOTE : All module types objects are removed, because otherwise it throws ValueError: Circular reference detected if not. TODO
+                if isinstance(x[i], ModuleType):
+                    del x[i]
+            return {'code': inspect.getsource(obj),
+                    'nonlocals': x}
+        elif isinstance(obj, np.ndarray):
+            return list(obj)
+        elif hasattr(obj, "__dict__"):
+            return getattr(obj, '__dict__')
+        elif isinstance(obj, np.uint8):
+            return int(obj)
+        try:
+            return json.JSONEncoder.default(self, obj)
+        except TypeError:
+            # This is used when the user enters an unknown type in CONFIG. Rather than throwing an error, we transform
+            # it into a string "Unsupported type for hashing" so it won't affect the hash.
+            return "Unsupported type for hashing"
+
+
+def get_json(object):
+    """Recursively serialize object to JSON. Use CustomEncoder class above.
```

Comment:
just a cosmetic change
```suggestion
    """Recursively serialize `object` to JSON using the :class:`CustomEncoder` class.
```